KAFKA-3338 [Kafka Streams] : Add print and writeAsText to KStream and KTable interfaces

bbejeck opened the pull request:
This is the first pass at implementing the print and writeAsText functionality for KStream and KTable.  I have some uncertainty on how I implemented the use of Serdes.

This contribution is my original work and I license this contribution to Apache-Kafka.

bbejeck committed:
KAFKA-3338: Initial pass at implementing print/writeAsText functionality for KStream and KTable.

bbejeck committed:
KAFKA-3338: Initial pass at implementing print/writeAsText functionality for KStream and KTable.

hachikuji committed:
KAFKA-2998: log warnings when client is disconnected from bootstrap brokers

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Grant Henke, Guozhang Wang

Closes #769 from hachikuji/KAFKA-2998

matthewlmcclure committed:
KAFKA-3384: Conform to POSIX kill usage

I believe this addresses KAFKA-3384.

The POSIX kill manpage is at http://pubs.opengroup.org/onlinepubs/9699919799/utilities/kill.html

Author: Matt McClure <mlm@aya.yale.edu>

Reviewers: Geoff Anderson <geoff@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1148 from matthewlmcclure/KAFKA-3384

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+
+        @Override
+        public void process(K key, V value) {
+            Object keyToPrint = maybeDeserialize(key, keySerde.deserializer());
```
These two lines will be changed to expect the correct types and if not correct a message more meaningful than ClassCastException will be given.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+        public void process(K key, V value) {
+            Object keyToPrint = maybeDeserialize(key, keySerde.deserializer());
+            Object valueToPrint = maybeDeserialize(value, valueSerde.deserializer());
+
```
These two lines (95, 96)  will be changed to expect the correct types.  If not a meaningful message will be provided (not just a ClassCastException)

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+                //TODO  calling this.processorContext.topic() throws Exception in unit test
+                return deserializer.deserialize("", (byte[]) receivedElement);
+            }
+
```
TODO will be removed, left to force the issue with myself that MockProcessorContext.topic() throws UnsupportedOperationException but I wanted to push the PR with unit tests. Should the MockProcessorContext.topic() method be updated to return default or empty string?

ijuma committed:
KAFKA-3510; OffsetIndex thread safety

* Make all fields accessed outside of a lock `volatile`
* Only allow mutation within the class
* Remove unnecessary `AtomicInteger` since mutation always happens inside a lock

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1188 from ijuma/kafka-3510-offset-index-thread-safety

ijuma committed:
KAFKA-3489; Update request metrics if a client closes a connection while the broker response is in flight

I also fixed a few issues in `SocketServerTest` and included a few clean-ups.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1172 from ijuma/kafka-3489-update-request-metrics-if-client-closes

granthenke committed:
KAFKA-3508: Fix transient SimpleACLAuthorizerTest failures

Allows the the maximum retires when writing to zookeeper to be overridden in tests and sets the value to Int.MaxValue to avoid transient failure.

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1156 from granthenke/transient-acl-test

mjsax committed:
KAFKA-3477: extended KStream/KTable API to specify custom partitioner for sinks

Author: mjsax <matthias@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1180 from mjsax/kafka-3477-streamPartitioner-DSL

bbejeck commented:
ping @guozhangwang please have a look if you can.

guozhangwang commented:
A general question: do we need to let print / write return the same object of KStream / KTable as well? For programmability, would users feel it more intuitive as:

```
stream1 = source.map(...).filter(...);
stream1.print();   // this is for debugging, deleted before moving to productiong

stream2 = stream1.transform(...);
stream2.print();   // this is for debugging, deleted before moving to productiong
```

Or

```
source.map(...).filter(...);
          .print();   // this is for debugging, deleted before moving to productiong
          .transform(...);
          .print();   // this is for debugging, deleted before moving to productiong
```

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -227,6 +232,41 @@ public void to(Serde<K> keySerde, Serde<V> valSerde, String topic) {
+
+        } catch (FileNotFoundException e) {
+            String message = "Unable to write stream to file at [" + filePath + "]";
+            throw new IllegalStateException(message, e);
```
Maybe return TopologyBuilderException since it will not be thrown in "runtime" after `KafkaStreams.start()` is triggered, but before that when the topology was built.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    KStreamPrinter(Serde<?> keySerde, Serde<?> valueSerde) {
+        this(null, keySerde, valueSerde);
```
Could we just return `System.out` instead of null here, and let the constructor in line 39 checking printStream not null?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+
+            String stringKey = keyToPrint == null ? NULL_KEY : keyToPrint.toString();
+
+            printStream.println(stringKey + " , " + valueToPrint);
```
Maybe also add the processor name as prefix in case there are more than one printer processor in the topology?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -131,6 +131,40 @@
     void to(Serde<K> keySerde, Serde<V> valSerde, String topic);
 
     /**
+     * Print the elements of this stream to System.out
```
We need to state that users are expected to override toString() function for these functions in JavaDoc. Similar comments apply below as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+        public void close() {
+            if (notStandardOut) {
+                this.printStream.close();
+            }
```
Do we need to call flush for System.out?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -155,6 +160,37 @@ public void to(Serde<K> keySerde, Serde<V> valSerde, String topic) {
+    @Override
+    public KTable<K, V> print(Serde<?> keySerde, Serde<?> valSerde) {
+        String name = topology.newName(PRINTING_NAME);
+        topology.addProcessor(name, new KStreamPrinter<>(keySerde, valSerde), this.name);
```
If we do not return the object (see the general comment), then we can just call corresponding KStreamImpl.xx functions.

guozhangwang commented:
Inviting @norwood to review as well.

norwood commented:
since streams is already depending on slf4j whats the motivation behind not just using slf4j for output instead of manually handling `System.out` and other `PrintStream`s.

you could have something like 

```
KStream<K, V> print(String name);
```

and then use that named logger to print that stream. this also allow you to externally config printing on/off so you don't even need to remove the `.print()` calls for production if you dont want to.

edit: this would also allow us to get rid of the `writeAsText` methods

norwood commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+import java.io.PrintStream;
+
+
+class KStreamPrinter<K, V> implements ProcessorSupplier<K, V> {
```
`KStreamPrinter` seems inaccurate since its also used with `KTable`s

norwood commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java:
```diff
@@ -0,0 +1,126 @@
+    private final PrintStream printStream;
+    private Serde<?> keySerde;
+    private Serde<?> valueSerde;
+    private static final String NULL_KEY = "null_key";
```
i believe that `bin/kafka-console-consumer.sh` will output `null` for a null key. we should probably match that.

enothereska committed:
KAFKA-3497: Streams ProcessorContext should support forward() based on child name

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Yuto Kawamura, Michael G. Noll, Guozhang Wang

Closes #1194 from enothereska/kafka-3497-forward

bbejeck commented:
@guozhangwang , @norwood 

I've addressed most of the comments, but I've screwed up and did a merge vs a rebase to get this branch current/remove conflict.  I don't want to push as it is now.  What are my options - kill the PR and resubmit with only my changes?

rajinisivaram committed:
KAFKA-3488; Avoid failing of unsent requests in consumer where possible

Fail unsent requests only when returning from KafkaConsumer.poll().

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1183 from rajinisivaram/KAFKA-3488

apovzner committed:
MINOR: ensure original use of prop_file in verifiable producer

This PR: https://github.com/apache/kafka/pull/958 fixed the use of prop_file in the situation when we have multiple producers (before, every producer will add to the config). However, it assumes that self.prop_file is initially "". This is correct for all existing tests, but it precludes us from extending verifiable producer and adding more properties to the producer config (same as console consumer).

This is a small PR to change the behavior to the original, but also make verifiable producer use prop_file method to be consistent with console consumer.

Also few more fixes to verifiable producer came up during the review:
-- fixed each_produced_at_least() method
-- more straightforward use of compression types

granders please review.

Author: Anna Povzner <anna@confluent.io>

Reviewers: Geoff Anderson <geoff@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1192 from apovzner/fix_verifiable_producer

bbejeck committed:
Kafka-3338 [Kafka Streams]
    Merged KStream to resolve conflict
    Changed print and writeAsText method return signature to void
    Flush System.out on close
    Changed IllegalStateException to TopologyBuilderException
    Updated MockProcessorContext.topic method to return a String
    Renamed KStreamPrinter to KeyValuePrinter
    Updated the printing of null keys to 'null' to match ConsoleConsumer
    Updated JavaDoc stating need to override toString

bbejeck committed:
Merge branches

# Conflicts:
#	streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
#	streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java

bbejeck commented:
@guozhangwang - 
       -      Changed print and writeAsText method return signature to void 
       -       Flush System.out on close, 
       -       Changed IllegalStateException to TopologyBuilderException, 
       -       Updated JavaDoc stating need to override toString, 

@norwood 
       -  Renamed KStreamPrinter to KeyValuePrinter
       -  Updated the printing of null keys to 'null' to match ConsoleConsumer

bbejeck commented:
> since streams is already depending on slf4j whats the motivation behind not just using slf4j for output instead of manually handling System.out and other PrintStreams.

@norwood , @guozhangwang 

What about keeping print() to System.out, but adding a print(Logger log) method that writes results to the given logger (log file)

guozhangwang commented:
@bbejeck @norwood The problem with using a given logger is that the logging level is universal, and hence users either 1) use a separate logger which is not used anywhere else, and set in log4j.properties the logging level for this specific logger, or 2) use a shared logger and tolerate it outputs other entries with the same logging level. Hence I'd prefer keep the current APIs and add the log4j support for print when we see common usage that requires it.

guozhangwang commented on bin/kafka-server-stop.sh:
```diff
@@ -19,6 +19,6 @@ if [ -z "$PIDS" ]; then
   echo "No kafka server to stop"
   exit 1
 else 
-  kill -SIGTERM $PIDS
```
Is these two changes intentional?

EDIT: nm, just realize it is due to rebasing issues.

guozhangwang commented:
@bbejeck As for the rebasing option, you can either rebase and squash the commits, or just create a new PR and close this one. Either is fine.

bbejeck commented:
@guozhangwang Thanks, I'll do the latter.

bbejeck commented:
Closing the PR due to merging/rebase issues.  Will re-open a new PR related to JIRA KAFKA-3338
